build/lib/src/library_cycle_graph/library_cycle_graph_loader.dart

```diff
@@ -39,27 +40,39 @@ import 'phased_value.dart';
 /// Secondly, because the loader is for use _during_ the build, it might be that
 /// not all files have been generated yet. So, results must be returned based on
 /// incomplete data, as needed.
+///
+/// There can be multiple concurrent computations running on top of the same
+/// state, as noted in the implementation. It is allowed to call the methods
+/// `libraryCycleOf`, `libraryCycleGraphOf` and `transitiveDepsOf` while a prior
+/// call to any of the methods is still running, provided each newer call is at
+/// an earlier phase. This happens when a load does a read that triggers a build
+/// of a generated file in an earlier phase.
 class LibraryCycleGraphLoader {
+  /// The phases at which evaluation is currently running.
+  ///
+  /// Used to check that recursive loads are always to an earlier phase.
+  final List<int> _runningAtPhases = [];
+
   /// The dependencies of loaded assets, as far as is known.
   ///
   /// Source files do not change during the build, so as soon as loaded
   /// their value is a [PhasedValue.fixed] that is valid for the whole build.
   ///
   /// A generated file that could not yet be loaded is a
   /// [PhasedValue.unavailable] specify the phase when it will be generated.
-  /// When to finish loading the asset is tracked in [_assetDepsToLoadByPhase].
+  /// When to finish loading the asset is tracked in [_idsToLoad].
   ///
   /// A generated file that _has_ been loaded is a [PhasedValue.generated]
   /// specifying both the phase it was generated at and its parsed dependencies.
   final Map<AssetId, PhasedValue<AssetDeps>> _assetDeps = {};
 
-  /// Generated assets that were loaded before they were generated.
+  /// Assets to load.
   ///
-  /// The `key` is the phase at which they have been generated and can be read.
-  final Map<int, Set<AssetId>> _assetDepsToLoadByPhase = {};
+  /// The `key` is the phase to load them at or after. A [SplayTreeMap] is used
+  /// to keep the keys sorted so earlier phases are processed first.
+  final SplayTreeMap<int, List<AssetId>> _idsToLoad = SplayTreeMap();
```

Comment:
Where is the sorted order used? I can't seem to find it.

Reply:
It's implicit in `_nextIdToLoad`, tweaked the comment to mention it.

Reply:
Yeah, it does `_idsToLoad.entries.where((entry) => entry.key <= upToPhase).first;` --- though wouldn't that just be a slow way to do `_idsToLoad.entries.first` then (well, and a check that i is in fact `<= upToPhase`)?